src/lower/lower_indexexpr_taco.cpp

```diff
@@ -0,0 +1,1168 @@
+
+    virtual void visit(const IndexedTensor* op) {
+      Expr src = op->tensor;
+      if (destIndices == op->indexVars) {
+        return;
+      }
+      if (isa<VarExpr>(dest)) {
+        if (isa<VarExpr>(src) &&
+            to<VarExpr>(dest)->var == to<VarExpr>(src)->var) {
+          value = true;
+        }
+      } else if (isa<TensorRead>(dest)) {
+        if (isa<TensorRead>(src) && getMainTensor(dest) == getMainTensor(src)) {
+          const TensorRead* d = to<TensorRead>(dest);
+          const TensorRead* s = to<TensorRead>(src);
+          if (!all_of(d->indices.begin(), d->indices.end(),
+                      (bool(&)(Expr))isa<Literal>) ||
+              !all_of(s->indices.begin(), s->indices.end(),
+                      (bool(&)(Expr))isa<Literal>) ||
+              !equal(d->indices.begin(), d->indices.end(), s->indices.begin(),
+                     [](const Expr d, const Expr s) {
+                       simit_iassert(isa<Literal>(d) && isa<Literal>(s));
+                       return *to<Literal>(d) == *to<Literal>(s);
+                     })) {
+            value = true;
+          }
+        }
+      } else if (isa<FieldRead>(dest)) {
+        if (isa<FieldRead>(src) && getMainTensor(dest) == getMainTensor(src) &&
+            to<FieldRead>(src)->fieldName == to<FieldRead>(dest)->fieldName) {
+          value = true;
+        }
+      }
+    }
+
+    virtual void visit(const AssignStmt* op) {
+      dest = op->var;
+      if (isa<IndexExpr>(op->value)) {
+        destIndices = to<IndexExpr>(op->value)->resultVars;
+        op->value.accept(this);
+      }
+    }
+
+    virtual void visit(const TensorWrite* op) {
+      dest = TensorRead::make(op->tensor, op->indices);
+      if (isa<IndexExpr>(op->value)) {
+        destIndices = to<IndexExpr>(op->value)->resultVars;
+        op->value.accept(this);
+      }
+    }
+
+    virtual void visit(const FieldWrite* op) {
+      dest = FieldRead::make(op->elementOrSet, op->fieldName);
+      if (isa<IndexExpr>(op->value)) {
+        destIndices = to<IndexExpr>(op->value)->resultVars;
+        op->value.accept(this);
+      }
+    }
+  } visitor;
+  stmt.accept(&visitor);
+  return visitor.value;
+}
+
+static vector<bool> generateSparsity(Type t) {
+  simit_iassert(t.isTensor());
+  const TensorType* type = t.toTensor();
+  vector<bool> sparsity; // true = sparse
+  flattenIndices(type->getDimensions(),
+      [](const IndexDomain& domain) { return domain.getNumIndexSets(); },
+      [](const IndexDomain& domain) { return domain.getIndexSets(); },
+      [&](const IndexSet& indexSet) {
+        sparsity.push_back(indexSet.getKind() != IndexSet::Range);
+      });
+  if (sparsity.size() > 0) {
+    sparsity[0] = false;
+  }
+  return sparsity;
+}
+
+// The same index var appears on different dimensions at different expressions,
+// while the sparsities of these dimensions do not agree
+// For example  A(i,j) = B(i,k)*B(k,j), where B is {Dense, Sparse}
```

Comment:
Is this actually an issue? taco seems to be able to generate code for the particular example in the comment.